clients/common/nm-vpn-helpers.c

```diff
@@ -247,3 +253,522 @@ nm_vpn_openconnect_authenticate_helper (const char *host,
+	}
+
+	if (!_wg_complete_peer (&data_peers,
+	                        g_steal_pointer (&current_peer),
+	                        current_peer_start_line_nr,
+	                        filename,
+	                        error))
+		return FALSE;
+
+	connection = nm_simple_connection_new ();
+	s_con = NM_SETTING_CONNECTION (nm_setting_connection_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_con));
+	s_ip4 = NM_SETTING_IP_CONFIG (nm_setting_ip4_config_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_ip4));
+	s_ip6 = NM_SETTING_IP_CONFIG (nm_setting_ip6_config_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_ip6));
+	s_wg = NM_SETTING_WIREGUARD (nm_setting_wireguard_new ());
+	nm_connection_add_setting (connection, NM_SETTING (s_wg));
+
+	uuid = nm_utils_uuid_generate ();
+
+	g_object_set (s_con,
+	              NM_SETTING_CONNECTION_ID,
+	              ifname,
+	              NM_SETTING_CONNECTION_UUID,
+	              uuid,
+	              NM_SETTING_CONNECTION_TYPE,
+	              NM_SETTING_WIREGUARD_SETTING_NAME,
+	              NM_SETTING_CONNECTION_INTERFACE_NAME,
+	              ifname,
+	              NULL);
+
+	g_object_set (s_wg,
+	              NM_SETTING_WIREGUARD_PRIVATE_KEY,
+	              data_private_key,
+	              NM_SETTING_WIREGUARD_LISTEN_PORT,
+	              data_listen_port,
+	              NM_SETTING_WIREGUARD_FWMARK,
+	              data_fwmark,
+	              NULL);
+
+	if (data_peers) {
+		for (i = 0; i < data_peers->len; i++)
+			nm_setting_wireguard_append_peer (s_wg, data_peers->pdata[i]);
+	}
+
+	for (is_v4 = 0; is_v4 < 2; is_v4++) {
+		const char *method_disabled = is_v4 ? NM_SETTING_IP4_CONFIG_METHOD_DISABLED : NM_SETTING_IP6_CONFIG_METHOD_IGNORE;
+		const char *method_manual   = is_v4 ? NM_SETTING_IP4_CONFIG_METHOD_MANUAL   : NM_SETTING_IP6_CONFIG_METHOD_MANUAL;
+		NMSettingIPConfig *s_ip     = is_v4 ? s_ip4                                 : s_ip6;
+		GPtrArray *data_dns         = is_v4 ? data_dns_v4                           : data_dns_v6;
+		GPtrArray *data_addr        = is_v4 ? data_addr_v4                          : data_addr_v6;
+
+		if (data_dns && !data_addr) {
+			/* When specifying "DNS", we also require an "Address" for the same address
+			 * family. That is because a NMSettingIPConfig cannot have @method_disabled
+			 * and DNS settings at the same time.
+			 *
+			 * We don't have addresses. Silently ignore the DNS setting. */
+			data_dns = NULL;
+		}
+
+		g_object_set (s_ip,
+		              NM_SETTING_IP_CONFIG_METHOD,
+		              data_addr ? method_manual : method_disabled,
+		              NULL);
+
+		if (data_addr) {
+			for (i = 0; i < data_addr->len; i++)
+				nm_setting_ip_config_add_address (s_ip, data_addr->pdata[i]);
+		}
+		if (data_dns) {
+			for (i = 0; i < data_dns->len; i++)
+				nm_setting_ip_config_add_dns (s_ip, data_dns->pdata[i]);
+		}
+
+		if (data_table == _TABLE_AUTO) {
+			/* TODO: not implemented. */
```

Comment:
In order for NetworkManager to actually be useful to people using WireGuard, you'll probably need to implement some variety of this. At least solve the general question: 0.0.0.0/0 is in AllowedIPs. You now know what they expect. How will you configure their routes? Policy routing? Explicit routes? Sky's the limit, but you have to do something. The table parameter, I think, is a particularly good way of doing it, but that's only one method. (<https://www.wireguard.com/netns/> describes a few.)

Reply:
On #305, the AllowedIPs ranges will be added as routes. So that part will be covered.

Handling of routing-loops ("/0") is a related but separate issue. Fixing that problem is not restricted to WireGuard but affects also "ip-tunnels" (e.g. vxlan). It needs a general solution in NetworkManager, which won't happen before 1.16.0.

The *(non-)* solution for the moment is to let the user solve this manually. For example,
- by adding static routes to the endpoints on (the profile of) the interface with the default-route. This can be done by configuring the other profile in NetworkManager, or via script.
- or by (manually) setting up suitable routing policy rules (via a script). Support for configuring rules is work in progress and will be added soon.

Reply:
> It needs a general solution in NetworkManager, which won't happen before 1.16.0.

Sounds like NM will not be useful for WireGuard, which is unfortunate. Hopefully GUIs don't arrive until 1.17, so that I don't have to deal with as many confused users on a regular basis or get in an unfortunate habit of telling people, "dont use NM; it's broken" or something. Certainly I don't want to start propagating alternative workarounds, since those tend to stick around, on blogs and in configs, after the deficiency is addressed and confuse people even more.

What's the purpose of putting this off for later?

> (e.g. vxlan)

People generally have different expectations of vxlan.

Reply:
> What's the purpose of putting this off for later?

That it takes consideration (time+effort) to get this right.

We'll just document NM's WireGuard support in 1.16.0 as experimental.

Reply:
Or just put the consideration in now before 1.16 drops? Policy routing isn't really that hard.